src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs

```diff
@@ -124,11 +127,76 @@ public ProjectAssemblyNameAction(string assemblyName)
+                }
+
+                public override TrackedGeneratorDriver TransformGeneratorDriver(TrackedGeneratorDriver generatorDriver)
+                {
+                    var generators = _analyzerReferences.SelectMany(a => a.GetGenerators()).ToImmutableArray();
+                    return new TrackedGeneratorDriver(generatorDriver.GeneratorDriver?.AddGenerators(generators));
+                }
+            }
+
+            internal sealed class RemoveAnalyzerReferencesAction : CompilationTranslationAction
+            {
+                private readonly ImmutableArray<AnalyzerReference> _analyzerReferences;
+
+                public RemoveAnalyzerReferencesAction(ImmutableArray<AnalyzerReference> analyzerReferences)
+                {
+                    _analyzerReferences = analyzerReferences;
+                }
+
+                public override TrackedGeneratorDriver TransformGeneratorDriver(TrackedGeneratorDriver generatorDriver)
+                {
+                    var generators = _analyzerReferences.SelectMany(a => a.GetGenerators()).ToImmutableArray();
+                    return new TrackedGeneratorDriver(generatorDriver.GeneratorDriver?.RemoveGenerators(generators));
+                }
+            }
+
+            internal sealed class AddAdditionalDocumentsAction : CompilationTranslationAction
+            {
+                private readonly ImmutableArray<TextDocumentState> _additionalDocuments;
+
+                public AddAdditionalDocumentsAction(ImmutableArray<TextDocumentState> additionalDocuments)
+                {
+                    _additionalDocuments = additionalDocuments;
+                }
+
+                public override TrackedGeneratorDriver TransformGeneratorDriver(TrackedGeneratorDriver generatorDriver)
+                {
+                    // PROTOTYPE: we need to track AdditionalTextWithState instances for later removal
+                    // return generatorDriver.WithPendingEdits(_additionalDocuments.SelectAsArray(a => (PendingEdit)new AdditionalFileAddedEdit(new AdditionalTextWithState(a))));
+                    return new TrackedGeneratorDriver(generatorDriver: null);
+                }
+            }
+
+            internal sealed class RemoveAdditionalDocumentsAction : CompilationTranslationAction
+            {
+                private readonly ImmutableArray<TextDocumentState> _additionalDocuments;
+
+                public RemoveAdditionalDocumentsAction(ImmutableArray<TextDocumentState> additionalDocuments)
+                {
+                    _additionalDocuments = additionalDocuments;
+                }
+
+                public override TrackedGeneratorDriver TransformGeneratorDriver(TrackedGeneratorDriver generatorDriver)
+                {
+                    // PROTOTYPE: right now we have no way to remove additional files so we have to do one from scratch
+                    return new TrackedGeneratorDriver(generatorDriver: null);
```

Comment:
not certain i understand hte comment.

Reply:
Reworded to be a bit clearer, hopefully?